Hooks/ilhooks.md

```diff
@@ -1,23 +1,369 @@
+IL_0056: ldarg.0
+>>>cursor is here<<< IL_005c: newobj instance void [UnityEngine.CoreModule]UnityEngine.WaitForSeconds::.ctor(float32)
+IL_0061: stfld object DropPlatform/'<Flip>d__16'::'<>2__current'
+```
+
+We can then add back a new value using `Emit`, restoring the stack to a valid state.
+
+```text
+// <>2__current = new WaitForSeconds(0.7f);
+IL_0056: ldarg.0
+>>>cursor is here<<< ldc.r4 SomeValue // this would be the actual constant value
+IL_005c: newobj instance void [UnityEngine.CoreModule]UnityEngine.WaitForSeconds::.ctor(float32)
+IL_0061: stfld object DropPlatform/'<Flip>d__16'::'<>2__current'
+```
+
+Here's the full code to do the above manipulations:
+
+```csharp
+if (cursor.TryGotoNext
+(
+    i => i.MatchLdcR4(0.7f),
+    i => i.MatchNewobj<WaitForSeconds>()
+))
+{
+    cursor.Remove();
+    cursor.Emit(OpCodes.Ldc_I4, SomeValue);
+}
+```
+
+You can emit any kind of opcode, with or without an operand. As a reminder, there is a full list of opcodes
+[here](https://en.wikipedia.org/wiki/List_of_CIL_instructions). You'll see most of the ones you'll usually use in
+the process of investigating the IL code you're hooking. Covering the functionality is out of scope for this
+documentation, but a few commonly used ones include:
+* ldc.i4
+* ldstr
+* ldarg
+* ldfld
+* callvirt
+
+You'll notice that most of these are ways to add values to the stack. In fact, if you want to do meaningful processing
+on the data, there is a much easier way to do that: by using `EmitDelegate`. This will allow you to emit an instruction
+that will call your C# code. This is a common place for the stack to get imbalanced, because arguments to your methods
+are removed from the stack. In general, this means you should do one of the following:
+* Use `EmitDelegate<Action>` to add additional code that needs no inputs and has no outputs.
+* Use `EmitDelegate<Func<T, T>>` to add additional code that takes an input and replaces it with an output of the
+  same type.
+* Add additional items to the stack, then use `EmitDelegate<Action<...>>` to add additional code that takes inputs
+  and has no outputs.
+* Remove items from the stack, then use `EmitDelegate<Func<T>>` to add code that does complex processing and then
+  adds a value to the stack.
+
+Of course, this is not a comprehensive list of all the possible ways you can handle inputs and outputs of delegates;
+they are just a few rules of thumb. Just make sure to keep the stack balanced.
+
+Let's see how to use `EmitDelegate` in action. Suppose we want to modify the drop platform time based on the player's
+current health. Here is one way to accomplish that using `EmitDelegate`.
+
+```csharp
+if (cursor.TryGotoNext
+(
+    i => i.MatchLdcR4(0.7f),
+    i => i.MatchNewobj<WaitForSeconds>()
+))
+{
+    // move between the ldc.r4 and the newobj; we want to have the original time on the stack as an input
+    cursor.GotoNext();
+    cursor.EmitDelegate<Func<float, float>>(f => {
```

Comment:
I assumed that this is not a programming in C# tutorial so I leave it to the reader to infer you can use a named function rather than a lambda. I like keeping it as a lambda here for brevity but can add a brief comment on this if needed.

Reply:
Brevity is more important so leave it lamda